New Continuous Integration section (#378)
crystal-lang/crystal-book: guides/ci/travis.md

```yml
# .travis.yml
language: bash

services:
  - docker

script:
  - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

**Note:** A list with the different official [Crystal docker images](https://hub.docker.com/r/crystallang/crystal/tags) is available at [DockerHub](https://hub.docker.com/r/crystallang/crystal).

### Using `latest`, `nightly` and a specific Crystal release all together!

For using 3 versions of the compiler we are going to use a different configuration file based on [Build Matrix](https://docs.travis-ci.com/user/customizing-the-build#build-matrix)
Here is the example:

```yaml
# .travis.yml
matrix:
 include:
   - language: crystal
     crystal:
       - latest
     script:
       - crystal spec

   - language: crystal
     crystal:
       - nightly
     script:
       - crystal spec

   - language: bash
     services:
       - docker
     script:
       - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

## Installing shards packages

Thanks to the `language: crystal` support, Travis CI automatically runs `shards install`. And for faster ci times we may use [caching](#caching)

> 

Reviewer: We should add:

When using a docker image you will need to run `shards install` explicitly. 

```
script:
  - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 shards install
  - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

Note: Since the shards will be installed in `./lib/` folder, it will be preserved for the second `docker run` command.

```yml
# .travis.yml
language: bash

services:
  - docker

script:
  - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

**Note:** A list with the different official [Crystal docker images](https://hub.docker.com/r/crystallang/crystal/tags) is available at [DockerHub](https://hub.docker.com/r/crystallang/crystal).

### Using `latest`, `nightly` and a specific Crystal release all together!

For using 3 versions of the compiler we are going to use a different configuration file based on [Build Matrix](https://docs.travis-ci.com/user/customizing-the-build#build-matrix)
Here is the example:

```yaml
# .travis.yml
matrix:
 include:
   - language: crystal
     crystal:
       - latest
     script:
       - crystal spec

   - language: crystal
     crystal:
       - nightly
     script:
       - crystal spec

   - language: bash
     services:
       - docker
     script:
       - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

> ## Using binary dependencies

Reviewer: Maybe we need to have a `## Installing shards dependencies` section where we said that 

Thanks to the `language: crystal` support, Travis CI automatically runs `shards install`. And reference `Caching` for faster ci times.

Then the binary dependencies section can focus only on that. Some shards require libraries and packages. You will need to install them using `apt` or similar.

Before presenting some examples, it is important to mention that there are many ways to achieve this and **it will heavily depend on our development workflow**.

With this in mind, let's continue!

Using shards dependencies (i.e. libraries declared in `shard.yml`) will not be a problem because, thanks to the `language: crystal` support, Travis CI automatically runs `shards install`.

But, how do we install dependencies outside `shards.yml`?
Well, here is a first example installing SQLite3 using Travis CI configuration file:

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

before_install:

>   - sudo apt-get -y install sqlite3

Reviewer: ```suggestion
  - sudo apt-get -y install libsqlite3-dev
```

this is the package that will normally be needed.

addons:
  apt:
    update: true

script:

>   - sqlite3 --version

Reviewer: ```suggestion
  - apt show libsqlite3-dev
```

and this is how we can check the package installed 🤷‍♂

  - crystal spec
```

> **Note:** We are using `sqlite3` just for printing SQLite's version, as an example of external dependency.

Reviewer: ```suggestion
**Note:** We are using `apt show` just for printing SQLite's version, as an example of binary dependency.
```

Reviewer 2: I think the idea was that `sqlite3` is installed as an executable. Of course we don't necessarily need that., using `libsqlite` is more precise.
But the wording needs to be changed: using `apt show` to display the installed package is not exactly an example of a binary dependency.

Author: Changing "using" with "installing"

The same can be accomplished using a containerization service like Docker. Here is an example of a `Dockerfile` installing SQLite3:

```dockerfile
# Dockerfile
FROM crystallang/crystal:latest

> RUN apt-get update && apt-get install -y sqlite3

Reviewer: ```suggestion
RUN apt-get update && apt-get install -y libsqlite3-dev
```

```

And here is the Travis CI configuration file:

```yml
# .travis.yml
language: bash

services:
  - docker

before_install:
  # build image using Dockerfile:
  - docker build -t testing .

script:
  # run specs in the container
  - docker run -v $PWD:/src -w /src testing crystal spec
  # sqlite3

>   - docker run -v $PWD:/src -w /src testing sqlite3 --version

Reviewer: ```suggestion
  - docker run -v $PWD:/src -w /src testing apt show libsqlite3-dev
```

#### Using Docker

When using a containerization service like Docker we will need to run `shards install` explicitly, like this:

```yml
# .travis.yml
language: bash

services:
  - docker

script:
  - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 shards install
  - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

**Note:** Since the shards will be installed in `./lib/` folder, it will be preserved for the second docker run command.

## Installing binary dependencies

> It is important to mention that there are many ways to achieve this and **it will heavily depend on our development workflow**.

Our application or maybe some shards may required libraries and packages. In that case we will need to install them using `apt` (or similar).

Here is a first example installing SQLite3 using Travis CI configuration file:

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

before_install:
  - sudo apt-get -y install libsqlite3-dev

addons:
  apt:
    update: true

script:
  - crystal spec
```

#### Using Docker

> The same can be accomplished using a containerization service like Docker. Here is an example of a `Dockerfile` installing SQLite3:

Reviewer 2: Using a `Dockerfile` is a very different method and introduces a completely new concept without much context. An example more similar to the previous section using `docker run` to install a dependency in the container is probably not a good idea, because it is more complex. But maybe that should be mentioned here? And a few more words about the method `Dockerfile`? When using a `Dockerfile` anyway, it would also make sense to install shards there.

However, a `Dockerfile` for setting up dependencies is not specific to travis CI, so maybe there should rather be a separate guide for more details which would be referenced here.

```dockerfile
# Dockerfile
FROM crystallang/crystal:latest

RUN apt-get update && apt-get install -y libsqlite3-dev
```

And here is the Travis CI configuration file:

```yml
# .travis.yml
language: bash

services:
  - docker

before_install:
  # build image using Dockerfile:
  - docker build -t testing .

script:
  # run specs in the container
  - docker run -v $PWD:/src -w /src testing crystal spec
```

**Note:** Dockerfile arguments can be used to use the same Dockerfile for latest, nightly or a specific version.

> ## Using a database (example using MySQL)

Reviewer 2: MySQL database is just an example, I'd call the section more general:
```suggestion
## Using services
```

> As you may notice, in the last configuration file we are using `docker` as a `service`. Travis CI may start different services as needed. For example, we may need [MySQL](https://docs.travis-ci.com/user/database-setup/#mysql).

Reviewer 2: I'm not sure the reference to the previous section is helpful. Readers who are only interested in travis runners have probably skipped it anyways. Maybe we could just swap dependencies and services section? This would avoid applying a concept before its introduction. Services like databases are probably more often used.

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

services:
  - mysql

script:
  - crystal spec
```

Here is the new test file for testing against the database:

```crystal
# spec/simple_db_spec.cr
require "./spec_helper"
require "mysql"

it "connects to the database" do
  DB.connect ENV["DATABASE_URL"] do |cnn|
    cnn.query_one("SELECT 'foo'", as: String).should eq "foo"
  end
end
```

When pushing this changes Travis CI will report the following error: `Unknown database 'test' (Exception)`, showing that we need to configure the MySQL service **and also setup the database**:

> > It's really **important** to notice that the lines we are adding to `.travis.yml` will depend exclusively on the development workflow we are using!

Reviewer 2: I don't understand this.

> > And remember that this is only an example using MySQL.

Reviewer 2: This should be mentioned in the first sentence of this section, and also a link to travis' service documentation.

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

env:
  - DATABASE_URL="mysql://root@localhost/test"

services:
  - mysql

before_install:
  - mysql -u root --password="" < test-data/setup.sql

script:
  - crystal spec
```

We are [using a `setup.sql` script](https://andidittrich.de/2017/06/travisci-setup-mysql-tablesdata-before-running-tests.html) to create a more readable `.travis.yml`. The file `./test-data/setup.sql` looks like this:

```sql

> -- setup.sql

Reviewer 2: The before_install step references this file at `test-data/setup.sql`. Maybe it should be `db/schema.sql` which is a very common path for this?

> CREATE DATABASE IF NOT EXISTS test;

Reviewer 2: Hardcoding the database name in the schema file is bad practice. It unnecessarily restricts the usage. For example, you can't have a production and test database on the same server.
Database creation should be handled with service setup so that shard finds the database at `DATABASE_URL` already existing.

-- CREATE TABLE ... etc ...
```

> Pushing these changes will trigger Travis CI and the build should be successful!!

Reviewer 2: ```suggestion
Pushing these changes will trigger Travis CI and the build should be successful.
```
Exclamation mark would be fine, but only a single one please 🙏

## Caching

If we read Travis CI job log, we will find that every time the job runs, Travis CI needs to fetch the libraries needed to run the application:

```log
Fetching https://github.com/crystal-lang/crystal-mysql.git
Fetching https://github.com/crystal-lang/crystal-db.git
```

This takes time and on the other hand these libraries might not change as often as our application, so it looks like we may cache them and save time.

> Travis CI [uses caching](https://docs.travis-ci.com/user/caching/) to improve some parts of the building path. For our example. Here is the new configuration file **with cache enabled**:

Reviewer 2: `For our example` - is some part of the sentence missing?

```yml
# .travis.yml
language: crystal
crystal:
  - latest

env:

>   - DATABASE_URL="mysql://root@localhost/test"

Reviewer 2: MySQL doesn't make sense here, it has nothing to do with caching.

cache:
  directories:

>     - $HOME/.cache/shards

Reviewer 2: `$HOME/.cache/shards` caches the git repos, but they still need to be checked out on every run. `$TRAVIS_BUILD_DIR/lib` should be cached as well. Or even only `$TRAVIS_BUILD_DIR/lib`.

Reviewer: I disagree. If the local lib directory is cached then it will not mimic a fresh checkout of the repo. The caching of the `.cache/shards` directory will prevent the down of the shards which should be the main source of time penalty.

In scenarios where a new version of the dependecy is released, caching the `$TRAVIS_BUILD_DIR/lib` will prevent checking the new release.

Reviewer 2: I guess you're right. When using a `shard.lock`, this shouldn't matter, though. But it's probably okay to cache only the shards cache, not checked out lib.

# Travis CI

If you are new to continuous integration (or you want to refresh the basic concepts) we may start reading the [core concepts guide](https://docs.travis-ci.com/user/for-beginners/) and then a [short tutorial](https://docs.travis-ci.com/user/tutorial/).

Now let's see an example of using Travis CI with Crystal.

## Build and run specs

### Using `latest` and `nightly`

When [creating a Crystal project](../../using_the_compiler/#creating-a-crystal-project) using `crystal init`, Crystal creates a `.travis.yml` file for us:

```yaml
# .travis.yml
language: crystal

crystal:
  - latest
  - nightly

script:
  - crystal spec
  - crystal tool format --check
```

Although a basic configuration file, it will let's use Travis CI. Now, let's go to Travis CI dashboard to [add the GitHub repository](https://docs.travis-ci.com/user/tutorial/#to-get-started-with-travis-ci). From this moment, Travis CI will run the `specs` against the Crystal compiler (using both `latest` and `nightly` releases)

### Using a specific Crystal release

> So, let's suppose we want to use version [0.31.1](https://github.com/crystal-lang/crystal/releases/tag/0.31.1)

Reviewer: Sometimes an app is unable to follow the upgrade of the underlying framework/language. It's better to have a way of not blocking the development of it.

 of the Crystal compiler. We will be tempting to add `- 0.31.1`. But, as we may read at the [Crystal configuration options](https://docs.travis-ci.com/user/languages/crystal/#configuration-options), we can only use the values `latest` and `nightly`.

So, how do we test against a specific version?

For this we are going to use [Docker](https://www.docker.com/). First we need to add Docker as a service in `.travis.yml`, and then use `docker` from the CLI to run the specs with a specific Crystal version, like this:

```yml
# .travis.yml
language: bash

services:
  - docker

script:
  - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

**Note:** We may find the list with the different official [Crystal docker images](https://hub.docker.com/r/crystallang/crystal/tags) at [DockerHub](https://hub.docker.com/r/crystallang/crystal).

> **Note:** This is a great trick that will let us use different Crystal versions without the need of actually installing the compiler and without the need of creating a `Dockerfile` configuration file.

Reviewer: Ok, we can remove the note

### Using `latest`, `nightly` and a specific Crystal release all together!

> For using 3 versions of the compiler we are going to use a different configuration file based on [Build Matrix](https://docs.travis-ci.com/user/customizing-the-build#build-matrix)

Reviewer: I see it as progressive support in Travis. It's better to use the native approach when possible.

Here is the example:

```yaml
# .travis.yml
matrix:
 include:
   - language: crystal
     crystal:
       - latest
     script:
       - crystal spec

   - language: crystal
     crystal:
       - nightly
     script:
       - crystal spec

   - language: bash
     services:
       - docker
     script:
       - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

> ## Using external dependencies

Reviewer 2: I would clarify this section. `shards` are also external dependencies, but not actually covered here. I suppose it's about binary dependencies?

Reviewer: Renaming it a binary dependencies sounds right.

Before presenting some examples, it is important to mention that there are many ways to achieve this and **it will heavily depend on our development workflow**.

With this in mind, let's continue!

> Using shards dependencies (i.e. libraries declared in `shard.yml`) will not be a problem, because thanks to the `language: crystal` support, TravisCI automatically runs `shards install`. But, how do we install dependencies outside `shards.yml`? Well, here is a first example installing SQLite3 using Travis CI configuration file:

Reviewer 2: The fact about shards being installed should definitely be mentioned earlier.
How does this work out with Docker, though? It wouldn't test installing shards with the shards version provided in the dockerfile, but the one in the travis runner.

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

before_install:
  - sudo apt-get -y install sqlite3

addons:
  apt:
    update: true

script:
  - sqlite3 --version
  - crystal spec
```

**Note:** We are using `sqlite3` just for printing SQLite's version, as an example of external dependency.

> The same can be accomplished using a containerization service like Docker. Here is an example of a `Dockerfile` installing SQLite3:

Reviewer 2: This is why I think running CI in Docker images should be separated topic. Also because it's not actually specific to travis (only the syntax for configuring it initially).

```

And here is the TravisCI configuration file:

```yml
# .travis.yml
language: bash

services:
  - docker

before_install:
  # build image using Dockerfile:
  - docker build -t testing .

script:
  # run specs in the container
  - docker run -v $PWD:/src -w /src testing crystal spec
  # sqlite3
  - docker run -v $PWD:/src -w /src testing sqlite3 --version
```

Again, these are only examples of two ways of running our application and installing (and using) dependencies.

> *Note:* Dockerfile arguments can be used to use the same Dockerfile for latest, nightly or a specific version.

## Using a database (example using MySQL)

As you may notice, in the last configuration file we are using `docker` as a `service`. Travis CI may start different services as needed. For example, we may need [MySQL](https://docs.travis-ci.com/user/database-setup/#mysql).

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

services:
  - mysql

script:
  - crystal spec
```

To continue this example, let's add a new test that uses the database.
(**important:** the new test case is only for testing Travis CI database service, and it's not a good design or test case by any means)

```crystal

> # game_of_life_spec.cr

Reviewer 2: This example is way to complex for a simple demonstration.
Just putting `DB.connect` in a spec file is completely sufficient.

Author: Yep, it is a little more complex example than `DB.connect` but it has another purpose: it shows how to use `environment variables` (in this case `DATABASE_URL`)
For making the example shorter, now we only show the lines that need to be added to the example introduced in Continuous Integration.

Reviewer 2: Really, I would just make the example be as tight as possible.
```cr
require "mysql"
require "spec"

it "connects do database" do
  DB.connect(ENV["DATABASE_URL"]) do |db|
    db.query("SELECT 'foo'", as: String). should eq "foo"
  end
end
```

There's no point in having anything else here. It's just noise.

Author: oh! ok, that's a simpler example, but I forgot to mention that the original example has another goal: to show how to use a `setup.sql` file to initialize the database from TravisCI. This may seem trivial but it's not (that's why I even mention the source of the solution: https://andidittrich.de/2017/06/travisci-setup-mysql-tablesdata-before-running-tests.html)
So we wanted to create an example that the reader could almost cut&paste.

I think that, after merging this new section, we may iterate with a new and more simple example (oh! and we would update the example in the introduction as well!) and then create a new PR with those changes. 

@straight-shoota, @bcardiff:  Does it seem ok?

By the way: thank you for all the suggestions!!

Reviewer 2: I seriously doubt that *setting up* a MySQL database fits in the scope of this guide. The goal of this section should be to demonstrate the concept of using services on travis.
How to actually use such a service is a highly specific topic. MySQL is really just a stand in for any travis service. You wouldn't want to dive into giving instructions for other services either.
It should be sufficient to refer to https://docs.travis-ci.com/user/database-setup/ for more specific documentation.

Author: ok! I removed the current example and added the simpler DB example.

Let me know what you think!

require "./spec_helper"

describe "a new world" do
  it "should be empty" do
    world = World.new
    world.is_empty?.should be_true
  end
end

describe "an empty world" do
  it "should not be empty after adding a cell" do
    world = World.empty
    world.set_living_at(Location.random)
    world.is_empty?.should be_false
  end

  it "should be persisted" do
    World.empty.save
  end
end
```

And the new implementation:

```crystal
# game_of_life.cr
require "mysql"

class Location
  getter x : Int32
  getter y : Int32

  def self.random
    Location.new(Random.rand(10), Random.rand(10))
  end

  def initialize(@x, @y)
  end
end

class World
  @living_cells : Array(Location)

  def self.empty
    World.new([] of Location)
  end

  def initialize(living_cells = [] of Location)
    @living_cells = living_cells
  end

  def set_living_at(a_location)
    @living_cells << a_location
  end

  def is_empty?
    @living_cells.size == 0
  end

  def save
    DB.connect ENV["DATABASE_URL"] do |cnn|
      cnn.exec("insert into worlds (living_cells) values (?);", @living_cells.size)
    end
  end
end
```

> We may use this `docker-compose` configuration to have a local MySQL database for our test to run:

Reviewer 2: This adds even more complexity and it doesn't even explain why you would need docker compose when you can install a mysql service in travis.

Reviewer: The locally running examples will go away and reduce the length.

language: crystal
crystal:
  - latest

env:
  - DATABASE_URL="mysql://root@localhost/test"

services:
  - mysql

before_install:
  - mysql -u root --password="" < test-data/setup.sql

script:
  - crystal spec
```

and then, the file `./test-data/setup.sql` will look like this:

```sql
-- setup.sql
CREATE DATABASE IF NOT EXISTS test;
CREATE TABLE `test`.`worlds` (`id` serial,`living_cells` int NOT NULL DEFAULT 0, PRIMARY KEY (id));
```

## Caching

If we read Travis CI job log, we will find that every time the job runs, Travis CI needs to fetch the libraries needed to run the application:

```log
Fetching https://github.com/crystal-lang/crystal-mysql.git
Fetching https://github.com/crystal-lang/crystal-db.git
```

This takes time and on the other hand these libraries might not change as often as our application, so it looks like we may cache them and save time.

Travis CI [uses caching](https://docs.travis-ci.com/user/caching/) to improve some parts of the building path. For our example. Here is the new configuration file **with cache enabled**:

```yml
# .travis.yml
language: crystal
crystal:
  - latest

env:
  - DATABASE_URL="mysql://root@localhost/test"

cache:
  directories:

>     - lib

Reviewer: ```suggestion
```

    - $HOME/.cache/shards

services:
  - mysql

before_install:
  - mysql -u root --password="" < travisci/test.sql

script:
  - crystal spec
```

> Let's push these changes. Travis CI will run, and it will install dependencies, but then it will cache the `lib/` and `$HOME/.cache/shards` folders. The following runs will use the cached dependencies.

Reviewer: ```suggestion
Let's push these changes. Travis CI will run, and it will install dependencies, but then it will cache the `$HOME/.cache/shards` folder. The following runs will use the cached dependencies.
```

# Travis CI

If you are new to continuous integration (or you want to refresh the basic concepts) we may start reading the [core concepts guide](https://docs.travis-ci.com/user/for-beginners/) and then a [short tutorial](https://docs.travis-ci.com/user/tutorial/).

Now let’s see an example of using Travis CI with Crystal.

## Build and run specs

### Using `latest` and `nightly`

> When [creating a Crystal project](https://crystal-lang.org/reference/using_the_compiler/#creating-a-crystal-project) using `crystal init`, Crystal creates a `.travis.yml` file for us:

Reviewer: Can we have relative links for thinks pointing to `/reference`?

```yaml
# .travis.yml
language: crystal

crystal:
  - latest
  - nightly

script:
  - crystal spec
  - crystal tool format --check
```

> Although a basic configuration file, it will let’s use Travis CI. Now, let’s go to Travis CI dashboard to [add the GitHub repository](https://docs.travis-ci.com/user/tutorial/#to-get-started-with-travis-ci) From this moment, Travis CI will run the `specs` against the Crystal compiler (using both `latest` and `nightly` releases)

Reviewer: ```suggestion
Although a basic configuration file, it will let’s use Travis CI. Now, let’s go to Travis CI dashboard to [add the GitHub repository](https://docs.travis-ci.com/user/tutorial/#to-get-started-with-travis-ci). From this moment, Travis CI will run the `specs` against the Crystal compiler (using both `latest` and `nightly` releases)
```

### Using a specific Crystal release

So, let’s suppose we want to use version [0.31.1](https://github.com/crystal-lang/crystal/releases/tag/0.31.1)
 of the Crystal compiler. We will be tempting to add `- 0.31.1`. But, as we may read at the [Crystal configuration options](https://docs.travis-ci.com/user/languages/crystal/#configuration-options), we can only use the values `latest` and `nightly`.

So, how do we test against a specific version?

For this we are going to use [Docker](https://www.docker.com/). First we need to add Docker as a service in `.travis.yml`, and then use `docker` from the CLI to run the specs with a specific Crystal version, like this:

```yml
# .travis.yml
language: bash

services:
  - docker

script:
  - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec spec/game_of_life_spec.cr
```

> **Note:** We may find the list with the different official [Crystal docker images](https://hub.docker.com/r/crystallang/crystal/tags) at [DockerHub](https://hub.docker.com/r/crystallang/crystal)

Reviewer: ```suggestion
**Note:** We may find the list with the different official [Crystal docker images](https://hub.docker.com/r/crystallang/crystal/tags) at [DockerHub](https://hub.docker.com/r/crystallang/crystal).
```

**Note:** This is a great trick that will let us use different Crystal versions without the need of actually installing the compiler and without the need of creating a `Dockerfile` configuration file.

### Using `latest`, `nightly` and a specific Crystal release all together!

For using 3 versions of the compiler we are going to use a different configuration file based on [Build Matrix](https://docs.travis-ci.com/user/customizing-the-build#build-matrix)
Here is the example:

```yaml
# .travis.yml
matrix:
 include:
   - language: crystal
     crystal:
       - latest
     script:
       - crystal spec

   - language: crystal
     crystal:
       - nightly
     script:
       - crystal spec

   - language: bash
     services:
       - docker
     script:

>        - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec spec/game_of_life_spec.cr

Reviewer: ```suggestion
       - docker run -v $PWD:/src -w /src crystallang/crystal:0.31.1 crystal spec
```

there is no need to use a different command `crystal spec` should work.

```

## Using external dependencies

> Before presenting some examples, it is important to mention that there are many ways to achieve this and **it will heavily depend on our development workflow**.

Reviewer: I think some intro is missing in this part. The goal is to install native libraries that might be linked by the app. `sqlite3` could be a better example than `git`. ( `sqlite3 --version` works ;-) )

And maybe a clarification, or short previous section, to mention that shards dependencies are installed directly in the `install` step thanks to the `language: crystal` support. Otherwise talking about dependencies could be misleading.

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

before_install:
  - sudo apt-get -y install git

addons:
  apt:
    update: true

script:
  - git --version
  - crystal spec spec/game_of_life_spec.cr
```

**Note:** We are using `git` just for printing GIT’s version, as an example of external dependency.

**Note:** Git is already installed by TravisCI and available but (again) we are using it as an example of installing and using a dependency.

The same can be accomplished using a containerization service like Docker. Here is an example of a `Dockerfile` installing Git:

```dockerfile
# Dockerfile
FROM crystallang/crystal:latest

RUN apt-get update && apt-get install -y git
```

And here is the TravisCI configuration file:

```yml
# .travis.yml
language: bash

services:
  - docker

before_install:
  # build image using Dockerfile:
  - docker build -t testing .
  # Git installed by Travis CI
  - git --version # => 2.21.0

script:
  # run specs in the container
  - docker run -v $PWD:/src -w /src testing crystal spec spec/game_of_life_spec.cr
  # Git installed in Docker container (see Dockerfile)
  - docker run -v $PWD:/src -w /src testing git --version # => 2.17.1
```

> We may notice that GIT’s version running in the `testing` container is different from Git running by default in Travis CI.

Reviewer: I would focus on the fact that the external dependency needs to be in the docker images.

> Again, these are only examples of two ways of running our application and installing (and using) dependencies.

Reviewer: I would add the following note.

> Note: Dockerfile arguments can be used to use the same Dockerfile for latest, nightly or a specific version.

There is no need for a full example.

## Using a database (example using MySQL)

> As you may notice, in the last configuration file we are using `docker` as a `service`. Travis CI may start different services as needed. For example, we may need  [MySQL](https://docs.travis-ci.com/user/database-setup/#mysql)

Reviewer: ```suggestion
As you may notice, in the last configuration file we are using `docker` as a `service`. Travis CI may start different services as needed. For example, we may need [MySQL](https://docs.travis-ci.com/user/database-setup/#mysql).
```

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

services:
  - mysql

script:
  - crystal spec
```

To continue this example, let’s add a new test that uses the database.
(**important:** the new test case is only for testing Travis CI database service, and it’s not a good design or test case by any means)

```crystal

> # game_of_life_spec.cr

Reviewer: ```suggestion
# spec/game_of_life_spec.cr
```

require "./spec_helper"

describe "a new world" do
  it "should be empty" do
    world = World.new
    world.is_empty?.should be_true
  end
end

describe "an empty world" do
  it "should not be empty after adding a cell" do
    world = World.empty
    world.set_living_at(Location.random)
    world.is_empty?.should be_false
  end

  it "should be persisted" do
    World.empty.save
  end
end
```

And the new implementation:

```crystal

> # game_of_life.cr

Reviewer: ```suggestion
# src/game_of_life.cr
```

require "mysql"

class Location
  getter x : Int32
  getter y : Int32

  def self.random
    Location.new(Random.rand(10), Random.rand(10))
  end

  def initialize(@x, @y)
  end
end

class World
  @living_cells : Array(Location)

  def self.empty
    World.new([] of Location)
  end

  def initialize
    initialize([] of Location)
  end

>   def initialize(living_cells)

Reviewer: ```suggestion
  def initialize(living_cells = [] of Location)
```

Use default args and the empty initialize can be removed.

    @living_cells = living_cells
  end

  def set_living_at(a_location)
    @living_cells << a_location
  end

  def is_empty?
    @living_cells.size == 0
  end

  def save

>     DB.connect ENV["DATABASE_URL"], do |cnn|

Reviewer: ```suggestion
    DB.connect ENV["DATABASE_URL"] do |cnn|
```

script:
  - crystal spec
```

Pushing these changes will trigger Travis CI and the build should be successful!!

We [use a `setup.sql` script](https://andidittrich.de/2017/06/travisci-setup-mysql-tablesdata-before-running-tests.html) to create a more readable `.travis.yml` file:

```yaml
# .travis.yml
language: crystal
crystal:
  - latest

env:
  - DATABASE_URL="mysql://root@localhost/test"

services:
  - mysql

before_install:
  - mysql -u root --password="" < test-data/setup.sql

script:
  - crystal spec
```

and then, the file `./test-data/setup.sql` will look like this:

```sql
-- setup.sql
CREATE DATABASE IF NOT EXISTS test;
CREATE TABLE `test`.`worlds` (`id` serial,`living_cells` int NOT NULL DEFAULT 0, PRIMARY KEY (id));
```

## Caching

If we read Travis CI job log, we will find that every time the job runs, Travis CI needs to fetch the libraries needed to run the application:

```log
Fetching https://github.com/crystal-lang/crystal-mysql.git
Fetching https://github.com/crystal-lang/crystal-db.git
```

This takes time and on the other hand these libraries might not change as often as our application, so it looks like we may cache them and save time.

> Travis CI [uses caching](https://docs.travis-ci.com/user/caching/) to improve some parts of the building path. For our example. here is the new configuration file **with cache enabled**:

Reviewer: ```suggestion
Travis CI [uses caching](https://docs.travis-ci.com/user/caching/) to improve some parts of the building path. For our example. Here is the new configuration file **with cache enabled**:
```

Reviewer: Someday this could be integrated directly in a `setup_cache` method at https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/script/crystal.rb . But for now, this section is needed.
